### PR TITLE
add min_acceptable_gain for better early stopping

### PR DESCRIPTION
**Pull request by robwec**
I run Optuna to optimize hyperparameters and often notice that it spends a long time "optimizing" for small loss gains, e.g. +/-0.0001. Therefore, it would make sense to stop early when the loss hasn't improved by at least some amount for N rounds. However, when I try to create a custom early stop callback imitating lightgbm's own early stop callback, it doesn't seem to play well with the exception for some reason. Therefore, editing the package source made the most sense.

This is a useful change. I can run grid search faster this way, since it will stop reasonably early rather than continuing 5000+ iterations in some cases just to squeeze out 0.0001 better loss on a lost case.

### CODE REVIEW

**Comment by robwec** [comment]
The question I was trying to answer when I made this: **How can I stop training early without a callback when loss doesn't increase by a specified amount (like say 0.001) for N rounds?**
